`StyleNano` removal: Palette system and Sheetlets

aspiringLich opened the pull request:
This PR is a continuation of a derelict PR, #28356. 
- [My changes from where the previous PR left off](https://github.com/space-wizards/space-station-14/compare/ac76c81062f499c351085364b2f996aea3c46a5c...aspiringLich:space-station-14:palette-gaming)
- [Previous PR's changes](https://github.com/moonheart08/space-station-14/compare/1777eea9a42a8bdbc4a865327bb9ef3e10fe57de..6b32bb2b148ef9d312b5332ccea4d8688694dba5)
- Overall changes you can find here, in the PR. Isn't that neat.
- [To-do list](https://github.com/aspiringLich/space-station-14/blob/palette-gaming/Content.Client/Stylesheets/Redux/todo.md) (ill delete before this is merged)
- [Design Doc I haven't touched](https://github.com/aspiringLich/space-station-14/blob/palette-gaming/Content.Client/Stylesheets/Redux/Stylesheets.md) (its out of date now i have a new better one)
- [codepen] for palette colors (cool)

## About the PR

### Summary 
- `StyleNano` is GONE
- Minimize hardcoding, colors are now much more consistent
- Centralized locations to define colors / resource paths / hardcoding goodness
- Actual documentation for UI dev, so people (like me) don't have to stare at UI code for 14 hours before doing anything (WIP).

## PROOF THE PR ACTUALLY WORKS: [link](https://imgur.com/a/JDXGNFA)

This won't break the game I pinky promise

## Why / Balance

https://github.com/space-wizards/space-station-14/blob/afc800274caaecf3647db2f04cd0635285eb675b/Content.Client/Stylesheets/StyleNano.cs#L42-L1611

## Technical details

This literally just rewrites content-side stylesheets from the ground up.

## Design Decisions

<details>
<summary><h3>This is really long so uh, click to expand</h3></summary><br />

As anyone who's had the misfortune of editing `StyleNano` is probably painfully aware of, the main problem with `StyleNano` is how difficult it is to find anything. Maybe the style classes you want are sitting somewhere in that gargantuan tangle of code, or maybe not. Maybe it;s just easier to hardcode everything in the UI code instead (WHICH IS THE ROUTE LOT OF PEOPLE HAVE (rightfully) TAKEN (stylenano was a complete fucking mess)). With this new iteration, I tried to maximize the readability of the *structure* of the code. Want to know what style classes are available for you to use? Look in `StyleClass`. Want to know the styles applicable to labels? `LabelSheetlet` or `TextSheetlet`. Buttons? `ButtonSheetlet`.

But what are `Sheetlet`s? Good question person-who-didnt-read-the-original-PR! Basically how styles used to work is every single style rule would be agglomerated in one unholy massive fucking list. This is still how it works, but now the responsibility of chipping in styles to this massive list is spread out among all the sheetlets. They have one method, `StyleRule[] GetRules`, and all these rules from all the sheetlets are collected up to do in like 30-40ish files what used to be one 1600 line file.

`ISheetletConfig` is intended to cut down on repeated code by providing shared functionality between the stylesheets. Its literally only used for buttons. Its also used (in my crusade against anything hardcoded) to store resource paths, because it's easier to reference if it's all centralized.

### Deviations from Original PR's Direction

[`StyleClass`]: https://github.com/aspiringLich/space-station-14/blob/215ad75077d4c1593d308c3d1ef3abcc05b61bf5/Content.Client/Stylesheets/Redux/StyleClass.cs
[`ColorPalette`]: https://github.com/aspiringLich/space-station-14/blob/215ad75077d4c1593d308c3d1ef3abcc05b61bf5/Content.Client/Stylesheets/Redux/ColorPalette.cs
[codepen]: https://codepen.io/aspiringLich/pen/VwOXdjd?editors=1000

The class holding all the style classes, [`StyleClass`] was originally named `Styleclasses`. I think this was a better name. Unfortunately, `Control`s have a field with the exact same name, meaning to reference `Styleclasses`, you would have to type out `Stylesheets.Redux.Styleclasses.<whatever>`. This syntax would be shortened after everything is moved out of `Redux` but still, annoying. The shorter syntax is nicer. Also as an extra bonus, it's closer to the syntax of something like `StyleClassLabelHeading`. Just add a dot! That's kinda neat.

<sub>class stopped sounding like a real word. Class class class class class. Blegh.</sub>

Instead of an array of colors, I made a custom [`ColorPalette`] class. This is because when transferring all the styles into `Sheetlet`s, I noticed I basically used the palette in one of three ways, foreground elements (Buttons etc.), background elements, (Panels, etc.), and text. So I represented that in the palette! Just makes code a bit more readable / robust. Also, I just kinda hated that `[0]` was the brightest and `[4]` the darkest. It should be the other way around! and what if I want to add more colors? There are more than five shades of any given color!

This does not compromise customizability because any colors can be curly-brace initialized after the fact if you are really motivated.

Also I made a cool shitty [codepen] to help me pick colors.

What moony originally did in their PR was they had resources scoped / access-locked (which I have kept), and, as a consequence, sheetlets that used a resource would have to be sheet-specific. This included buttons, panels, windows, etc.; things that basically every stylesheet would want, and which would have to be duplicated for every stylesheet. I think the intention behind this is to prevent resources intended for one stylesheet being used unintentionally in another, but honestly, the organization of the resources folder is a mess (a lot of style-generic resources are sitting in `Textures/Interface/Nano`), and this would lead to a lot of copying resources, which is definitely a maintenance hazard. I could clean it up, but then it could be a NIGHTMARE to merge for downstream forks (probably idk) (also touching the resource folder scares me). The resource-scoping system does allow you to specify multiple scopes to try in order, but that's kinda icky. So! I propose (and have already implemented) the following solution:

#### `GetResourceOr`

<sub>(and `GetTextureOr`)</sub>

It gets the scoped resource like normal, but if it doesn't exist, falls back to an absolute root. Is it a generic resource that has no business being is `Textures/Interface/Nano`? I don't care! It can stay there! I really can't be bothered to move it! Now the sheetlet will work with any stylesheet. If the scope happens to exist, it'll use that instead.

This method is kinda clunky but that's good (maybe?) because you want to use it where it counts. If there's a resource that, ideally, every theme should have a unique implementation of, then you would just use `GetResource`.

### Significant Interface Changes

- Colors are different (duh)
- Tooltips on action menu now is slightly transparent for consistency w/ examine popup
- Everything else should pretty much be the same. I'm like 80% sure I transferred everything from `StyleNano` correctly, as I was pretty methodical with it, but its also entirely possible I forgot something. If anybody notices UI that looks worse than they remember, double check it actually is different (this happened to me several times, some of the UIs are just kinda bad), and pretty please PR the changes (or ask me, and I'll probably do it).
- I definitely like messed a couple colors up but whatever
- Also anything that previously used `StyleSpace` now uses `SheetSystem` which is like `StyleSpace` but the colors are slightly different

</details>

## Breaking Changes

#### `StyleNano` is gone!!

- Style classes defined in `StyleNano` have been moved either to `Content.Client.Stylesheets.Redux.StyleClass` or their associated `Control`.
- Any `StyleRule` additions to `StyleNano` must be rewritten to conform to the new format (See guide **(TODO: LINK GUIDE)**)
- Some unused / redundant style classes were removed. If you hapenned to be relying on them, either substitute in the appropriate style classes defined in `Content.Client.Stylesheets.Redux.StyleClass` or create a new Sheetlet (Again, see guide)
- If you are planning on creating a new UI, there are new conventions you should follow (Yet again, see guide)
- `StyleSpace` is gone too but that probably doesn't affect you

#### Also

`square.png` has been deleted from `Textures/Interface/Nano`. This shouldn't have even been used in the first place `StyleBoxFlat` exists.

## Future work
- Allow the user to change the colors
- Allow the user to change the primary font.
- Syndicate/domain specific stylesheets (because they're no longer a major maintenance hazard)
- Reorganization; At this time all new tools, sheets, code, etc are in` Content.Client/Stylesheets/Redux` for ease of review / merging
- There's still a lot of `TODO`s left scattered around for hardcoding and stuff
- Also palettize margins and stuff?
- kill `DefaultWindow`
- Change all admin / sandbox windows to the new theme

**Changelog**

:cl:
- tweak: UI colors have been changed (new palette system)

aspiringLich commented:
This PR should not be merged until I have written documentation. For now, I am going to bed.

aspiringLich commented:
Oh whoops forgot to commit changes.

Deserty0 commented:
you are my hero

chavonadelal commented:
legendary

mirrorcult commented:
unfathomably blessed PR

beck-thompson commented:
I don't even want to think how much time this took 😆

LankLTE commented:
If you're going to continue a derelict PR please give CL and commit credit to the original author.

aspiringLich commented:
Oh yeah I should do that.

Also sidenote: I will be busy for the next while so documentation will have to wait a bit. I will write it eventually I didn't spend 3 weeks on this PR for nuthin'.

aspiringLich commented:
I am now free and writing documentation. If someone reviewed the PR that would be kinda cool, I think.

aspiringLich commented:
![image](https://github.com/user-attachments/assets/bcdf1f37-79dc-4847-9219-4ac128704cb5)

aspiringLich commented:
oh that uh, wasnt supposed to happen. hm.

aspiringLich commented:
Okay, I may need some help to help give Moony commmit credit. I'm decently familiar with git, but I'm not a wizard and this is tricky.

I tried checking out the commit in question, amending it with `git commit --amend --author` to change the author to moony, and rebasing it back onto the original commit, but that messed everything up because now this PR suddenly had like 1000 commits as github didnt realize those commits werent mine I guess.

Not sure what to do now, that was the one idea I had.

aspiringLich commented:
I'll resolve the merge conflicts once someone gives me the go ahead because _insert Sisyphus image_

PJB3005 commented on Content.Client/Stylesheets/Redux/SheetletConfigs/IButtonConfig.cs:
```diff
@@ -0,0 +1,89 @@
+    public ColorPalette PositiveButtonPalette { get; }
+    public ColorPalette NegativeButtonPalette { get; }
+
+    public StyleBox ConfigureBaseButton(IStyleResources sheet)
```
Yeah this really is not a good use of DIMs. I would rather these be extension methods or something.

aspiringLich replied:
These should be on `ButtonSheetlet` anyway probably

PJB3005 commented on Content.Client/Stylesheets/Redux/SheetletConfigs/IPanelConfig'.cs:
This file name is messed up.

PJB3005 commented on Content.Client/Stylesheets/Redux/Sheetlets/ButtonSheetlet.cs:
```diff
@@ -0,0 +1,103 @@
+        return rules.ToArray();
+    }
+
+    public static void MakeButtonRules<T>(IButtonConfig _,
```
Why take this `IButtonConfig` parameter if you're not using it?

PJB3005 commented on Content.Client/Stylesheets/Redux/Sheetlets/ButtonSheetlet.cs:
```diff
@@ -0,0 +1,103 @@
+        });
+    }
+
+    public static void MakeButtonRules(IButtonConfig _,
```
Ditto, unused parameter.

PJB3005 reviewed (changes requested):
didn't do a whole review but this should be the core of the actual system + half of the rest.

aspiringLich commented on Content.Client/Stylesheets/Redux/StyleClass.cs:
```diff
@@ -0,0 +1,60 @@
+    public const string LabelKeyText = "LabelKeyText";
+    public const string LabelWeak = "LabelWeak"; // replaces `StyleClassLabelSecondaryColor`
+
+    public const string BackgroundPanel = "AngleRect";
```
holdover from ass naming conventions in `StyleNano`. I'll rename it for consistency after resolve merge conflicts.

aspiringLich commented on Content.Client/Ame/UI/AmeWindow.xaml.cs:
```diff
@@ -32,9 +32,9 @@ public void UpdateState(BoundUserInterfaceState state)
             // Disable all buttons if not powered
-            if (Contents.Children.Any())
+            if (ContentsContainer.Children.Any())
             {
-                ButtonHelpers.SetButtonDisabledRecursive(Contents, !castState.HasPower);
+                ButtonHelpers.SetButtonDisabledRecursive(ContentsContainer, !castState.HasPower);
```
i changed it while warning hunting / cleaning up so, no idea! reverting.

aspiringLich commented on Content.Client/ContextMenu/UI/ContextMenuElement.xaml:
```diff
@@ -13,7 +13,8 @@
                 HorizontalAlignment="Right"
                 VerticalAlignment="Bottom"
-                StyleClasses="contextMenuIconLabel"
+                StyleClasses="contextMenuIconLabel LabelSubText"
```
oh rip

aspiringLich replied:
wait actually neither `contextMenuIconLabel` nor `StyleClassEntityMenuIconLabel` ever get used so I can just remove it

aspiringLich commented on Content.Client/Lobby/UI/CharacterSetupGui.xaml:
```diff
@@ -22,19 +22,16 @@
-                        <gfx:StyleBoxFlat BackgroundColor="{x:Static style:StyleNano.NanoGold}" ContentMarginTopOverride="2" />
-                    </PanelContainer.PanelOverride>
+                <PanelContainer MinSize="2 0" StyleClasses="highlight">
+                    <!-- <PanelContainer.PanelOverride> -->
+                    <!--     <gfx:StyleBoxFlat BackgroundColor="highlight" ContentMarginTopOverride="2" /> -->
+                    <!-- </PanelContainer.PanelOverride> -->
```
o7

aspiringLich commented on Content.Client/Stylesheets/IStylesheetManager.cs:
```diff
@@ -1,12 +1,14 @@
+    Stylesheet SheetNanotransen { get; }
+    Stylesheet SheetSystem { get; }
+
+    public Dictionary<string, Stylesheet> Stylesheets { get; }
```
i do it as a dictionary because `FancyWindow` has a `Stylesheet` property and needs to access a sheet by name, i could put that into a helper method maybe?

PJB3005 replied:
Yes, do that.

aspiringLich commented on Content.Client/Stylesheets/IStylesheetManager.cs:
```diff
@@ -1,12 +1,14 @@
-    public interface IStylesheetManager
-    {
-        Stylesheet SheetNano { get; }
-        Stylesheet SheetSpace { get; }
+    Stylesheet SheetNanotransen { get; }
+    Stylesheet SheetSystem { get; }
```
I've thought about this for a bit and I think it would also be worthwhile to keep `StyleNano.cs` and `StyleSpace.cs` as they were, but just obselete them. Makes merging easier.

aspiringLich commented on Content.Client/Stylesheets/Redux/Colorspace/OkLab.cs:
```diff
@@ -0,0 +1,106 @@
+///     Oklch hue and chroma are also provided.
+/// </summary>
+[PublicAPI]
+public struct OklabColor
```
oh yeah. Huh. there should be a warning for that

aspiringLich commented on Content.Client/Stylesheets/Redux/Colorspace/OkLab.cs:
```diff
@@ -0,0 +1,106 @@
+
+    public OklabColor(Color c)
+    {
+        // I won't pretend to know how this works.
+        // https://bottosson.github.io/posts/oklab/
+        var l = 0.4122214708d * c.R + 0.5363325363d * c.G + 0.0514459929d * c.B;
+        var m = 0.2119034982d * c.R + 0.6806995451d * c.G + 0.1073969566d * c.B;
+        var s = 0.0883024619d * c.R + 0.2817188376d * c.G + 0.6299787005d * c.B;
+
+        // ReSharper disable InconsistentNaming
+        var l_ = double.Cbrt(l);
+        var m_ = double.Cbrt(m);
+        var s_ = double.Cbrt(s);
+        // ReSharper restore InconsistentNaming
+
+        L = (float) (0.2104542553d * l_ + 0.7936177850d * m_ - 0.0040720468d * s_);
+        A = (float) (1.9779984951d * l_ - 2.4285922050d * m_ + 0.4505937099d * s_);
+        B = (float) (0.0259040371d * l_ + 0.7827717662d * m_ - 0.8086757660d * s_);
+        Alpha = c.A;
+    }
+
+    public static explicit operator Color(OklabColor c)
+    {
+        // ReSharper disable InconsistentNaming
+        var l_ = c.L + 0.3963377774d * c.A + 0.2158037573d * c.B;
+        var m_ = c.L - 0.1055613458d * c.A - 0.0638541728d * c.B;
+        var s_ = c.L - 0.0894841775d * c.A - 1.2914855480d * c.B;
+        // ReSharper restore InconsistentNaming
+
+        var l = l_ * l_ * l_;
+        var m = m_ * m_ * m_;
+        var s = s_ * s_ * s_;
+
+        return new(
+            (float) (+4.0767416621d * l - 3.3077115913d * m + 0.2309699292d * s),
+            (float) (-1.2684380046d * l + 2.6097574011d * m - 0.3413193965d * s),
+            (float) (-0.0041960863d * l - 0.7034186147d * m + 1.7076147010d * s),
+            c.Alpha
+        );
+    }
```
mm, moony wrote this part and i didnt catch that
ill just translate over https://gist.github.com/earthbound19/e7fe15fdf8ca3ef814750a61bc75b5ce which seems to be correct

aspiringLich commented on Content.Client/Stylesheets/Redux/Fonts/FontStack.cs:
```diff
@@ -0,0 +1,90 @@
+    /// </summary>
+    /// <param name="kind">Which font kind to use.</param>
+    /// <returns>An array of </returns>
+    protected virtual string[] GetFontPaths(FontKind kind)
+    {
+        if (!AvailableKinds.Contains(kind))
+        {
+            if (kind == FontKind.BoldItalic && AvailableKinds.Contains(FontKind.Bold))
+            {
+                kind = FontKind.Bold;
+            }
+            else
+            {
+                kind = FontKind.Regular;
+            }
+        }
+
+        var simpleKindStr = kind.SimplifyCompound().AsFileName();
+        var boldOrRegularStr = kind.RegularOr(FontKind.Bold).AsFileName();
+
+        var kindStr = kind.AsFileName();
+        var fontList = new List<string>()
+        {
+            string.Format(FontPrimary, kindStr, simpleKindStr, boldOrRegularStr),
+            string.Format(FontSymbols, kindStr, simpleKindStr, boldOrRegularStr),
+        };
+        fontList.AddRange(Extra);
+        fontList.Add(FontFallback);
+        return fontList.ToArray();
+    }
```
I think this was just pretty much copied over verbatim from `StyleNano.cs` and moony probably intended to go back and rework it at some point. 

I can alternatively rework it to specify different paths for each font face, or you can do the engine PR. Up to you.

aspiringLich replied:
leaving all font-related comments unresolved for now

PJB3005 replied:
If this is just a holdover from `StyleNano` then I'd say this definitely shouldn't have been moved into a "general purpose" looking thing like this. If this can be moved into `NotoFontStack` and the base type being left as an "uuhhh" for now, that's good enough IMO.

aspiringLich commented on Content.Client/Stylesheets/Redux/BaseStylesheet.Sheetlets.cs:
```diff
@@ -0,0 +1,50 @@
+
+public abstract partial class BaseStylesheet
+{
+    public StyleRule[] GetSheetletRules<TSheetTy>(Type sheetletTy)
```
maybe it could be given a better name, but `[CommonSheetlet]` makes the sheetlet accessible to the stylesheets, but all the valid ones should be loaded by the stylesheets. if it is not loaded it is because it does not meet the type constaints. 

That's a good idea though, I'll add a check that will log any `[CommonSheetlet]`s that are not loaded by any stylesheet.

aspiringLich commented on Content.Client/Stylesheets/Redux/BaseStylesheet.Sheetlets.cs:
```diff
@@ -0,0 +1,50 @@
+        }
+        // thrown when `sheetletTy.MakeGenericType` is given a type that does not satisfy the type constraints of
+        // `sheetletTy`
+        catch (ArgumentException) { }
```
yeah, this is the cleanest solution i've found without doing that. Tested a couple other methods and this ended up being the least slow. It's 170ms right now, so this probably should be fixed otherwise it could easily add a second or two to load times in the future.

I'm really not familliar with sandboxing / reflection, so a couple pointers would be helpful if I were to try and implement this.

aspiringLich commented on Content.Client/Paper/UI/StampWidget.xaml:
```diff
@@ -5,6 +5,6 @@
 
     <paper:StampLabel Name="StampedByLabel" StyleClasses="LabelHeadingBigger"
-        FontColorOverride="{x:Static style:StyleNano.DangerousRedFore}"
+        FontColorOverride="#cf2f2f"
```
This is for stamp labels, which get their color overridden anyway, so I'm not sure why this had to be set in the first place. Seems to work when removed.

aspiringLich commented:
Most of the trivial changes have been dealt with, waiting on comments. Also updated the docs PR.

I'll finish this up and merge upstream changes tomorrow.

aspiringLich commented:
Okay, I've spent a while going through every single UI I can think of and double-checking every single little thing and everything looks good I think? I caught a few stragglers (as you can see in the commits)

So, fonts. I think adding the engine support for exposing font weight / slant / whatever is a good idea. Currently, I took your suggestion and left `NotoFontFamilyStack` in there temporarily, and marked it as temporary.

Stuff like font sizes and padding and whatnot is also hard-coded. It's probably fine for now, a lot less terrible than what was hapenning with colors, but this should be standardized at some point.

> This definitely should be done with an explicit check instead instead of relying on exceptions to be thrown. That said, I think some of the reflection APIs you'd need are missing from sandbox so you'd need to add those

It currently isn't a dealbreaker honestly, the current system adds a tenth of a second to load times, but I am probably not familiar enough with sandboxing / c# dark magic to implement it properly, so, leaving that unresolved for now.

That should be pretty much all the review dealt with. I thought of a cleaner way of doing what I'm doing with `SheetletConfigs` with dependency injection, but that should just be refactoring. Unless you have any final comments, it should pretty much be in good shape to merge (after I do that thing).

Anyway, final steps (assuming another round of review doesn't come in):

*   Move everything out of `Redux`
*   Delete `todo.md` and `Stylesheets.md`
*   Make github issues about two unfinished things mentioned above
*   Get a final list of all the changed style classes and stuff for downstream forks

If ye have any final comments, say them now

aspiringLich commented:
Okay, I've spent a while going through every single UI I can think of and double-checking every single little thing and everything looks good I think? I caught a few stragglers (as you can see in the commits)

So, fonts. I think adding the engine support for exposing font weight / slant / whatever is a good idea. Currently, I took your suggestion and left `NotoFontFamilyStack` in there temporarily, and marked it as temporary.

Stuff like font sizes and padding and whatnot is also hard-coded. It's probably fine for now, a lot less terrible than what was hapenning with colors, but this should be standardized at some point.

> This definitely should be done with an explicit check instead instead of relying on exceptions to be thrown. That said, I think some of the reflection APIs you'd need are missing from sandbox so you'd need to add those

It currently isn't a dealbreaker honestly, the current system adds a tenth of a second to load times, but I am probably not familiar enough with sandboxing / c# dark magic to implement it properly, so, leaving that unresolved for now.

That should be pretty much all the review dealt with. I thought of a cleaner way of doing what I'm doing with `SheetletConfigs` with dependency injection, but that should just be refactoring. Unless you have any final comments, it should pretty much be in good shape to merge (after I do that thing).

Anyway, final steps (assuming another round of review doesn't come in):

*   Move everything out of `Redux`
*   Delete `todo.md` and `Stylesheets.md`
*   Make github issues about two unfinished things mentioned above
*   Get a final list of all the changed style classes and stuff for downstream forks

If ye have any final comments, say them now

chavonadelal commented:
When 😭

aspiringLich commented:
next week (tm)

chavonadelal commented:
> next week (tm)

I'll wait hehehe

Tornado-Technology commented:
Wheeeeen... I don't like nextweek™

chavonadelal commented:
😭

aspiringLich commented:
I dislike review limbo as much as you guys.

@PJB3005 the people hunger.

PJB3005 commented:
I did start reviewing it today, I promise.

PJB3005 reviewed (commented):
Yeah it's mostly good tbh we'll fix the rest in post

eoineoineoin commented on Content.Client/Stylesheets/IStylesheetManager.cs:
```diff
@@ -1,12 +1,32 @@
-        Stylesheet SheetNano { get; }
-        Stylesheet SheetSpace { get; }
+    /// Nanotransen styles: the default style! Use this for most UIs
+    Stylesheet SheetNanotransen { get; }
```
"Nanotransen" typo - also appears in a handful of other places.

aspiringLich replied:
I honestly fully believed thats how you spelt Nanotrasen

eoineoineoin commented:
I appreciate this is subjective, but I find the automatic pallete for a disabled `Button` is hard to read. Compare the before and after:

![image](https://github.com/user-attachments/assets/06c32dbb-f1bd-4c10-8b68-577d30f8eb91)

The borders of those buttons are very hard to distinguish from the background they're on - I can't see where "Copy" ends and "Send" begins, for example.

The `Button`'s child Label text is no longer modulated when the button is disabled. That behaviour should be restored, as the bright text is making it less obvious that the button is disabled.

aspiringLich commented:
Hm, honestly looking by them side by side too, I like the previous colors of the disabled button better too. I'll definitely tweak this. And do all the other review as soon as I can haha.

aspiringLich commented on Content.Client/ContextMenu/UI/ContextMenuElement.xaml:
```diff
@@ -13,7 +13,8 @@
                 Name="IconLabel"
                 HorizontalAlignment="Right"
                 VerticalAlignment="Bottom"
-                StyleClasses="contextMenuIconLabel"
```
Logic is: style classes should be reusable between multiple elements. If you want specific styling, use an style identifier and a specific sheetlet. In this case, it was just acheivable with styleclasses

aspiringLich commented on Content.Client/Lobby/UI/HumanoidProfileEditor.xaml.cs:
```diff
@@ -527,7 +528,7 @@ public void RefreshTraits()
                         Text = Loc.GetString(category.Name),
                         Margin = new Thickness(0, 10, 0, 0),
-                        StyleClasses = { StyleBase.StyleClassLabelHeading },
+                        StyleClasses = { StyleClass.LabelSubText },
```
oops

aspiringLich commented on Content.Client/MassMedia/Ui/NewsArticleCard.xaml:
```diff
@@ -17,13 +17,13 @@
             <Control HorizontalExpand="True"/>
             <Label FontColorOverride="#b1b1b2" StyleClasses="LabelSmall" Name="PublishTimeLabel" Margin="6 6 6 6"/>
-            <controls:ConfirmButton Name="DeleteButton" Text="{Loc news-write-ui-delete-text}"
+            <Button Name="DeleteButton" Text="{Loc news-write-ui-delete-text}"
```
no idea

aspiringLich commented on Content.Client/Medical/CrewMonitoring/CrewMonitoringWindow.xaml.cs:
```diff
@@ -111,7 +112,6 @@ public void ShowSensors(List<SuitSensorStatus> sensors, EntityUid monitor, Entit
             };
 
             deparmentLabel.SetMessage(department);
-            deparmentLabel.StyleClasses.Add(StyleNano.StyleClassTooltipActionDescription);
```
oh huh I guessed I missed it while pruning stuff out.

as another note currently `ActionTooltip` contains classes that I'm fairly certain never did anything since the style class was called `tooltipActionTitle` and not `StyleClassTooltipActionTitle`, so I'm leaving a comment noting that.

aspiringLich replied:
oh `ActionTooltip` is just never used anywhere. That explains it.

aspiringLich commented on Content.Client/Power/PowerCharge/PowerChargeWindow.xaml.cs:
```diff
@@ -41,7 +43,10 @@ public void UpdateState(PowerChargeState state)
             ("max", state.PowerDrawMax));
 
-        PowerLabel.SetOnlyStyleClass(MathHelper.CloseTo(state.PowerDraw, state.PowerDrawMax) ? "Good" : "Caution");
+        PowerLabel.FontColorOverride = MathHelper.CloseTo(state.PowerDraw, state.PowerDrawMax)
+            ? Palettes.Status.GetStatusColor(1.0f)
+            : Palettes.Status.GetStatusColor(0.5f);
+        // TODO: blend with status colors instead of setting hard values
```
I added some style classes and color constants but color constants probably work better here /shrug

aspiringLich commented:
![](https://github.com/user-attachments/assets/a27327ea-0626-4da0-a89b-0026e73fae37)

Colors have been tweaked, now resembles original more.

aspiringLich commented on Content.Client/Stylesheets/Redux/Fonts/FontKind.cs:
```diff
@@ -0,0 +1,12 @@
+/// <summary>
+///     The available kinds of font.
+/// </summary>
+public enum FontKind
+{
+    Regular,
+    Bold,
+    Italic,
+    BoldItalic,
+}
```
resolve: awaiting engine font support

aspiringLich replied:
resolve: awaiting engine font support

aspiringLich commented:
Okay! So, I'm pretty sure that should be done. I moved everything out of `Redux`, nothing major changed, so we should be fine. Updated the breaking changes, we should be pretty much good to merge...

Also, is the size labeler broken? This PR is a large pr, but its being labeled as a medium size one.

aspiringLich commented:
PJB has the better sandboxing approach she said that she would push to my branch and also engine font support to add because currently the font implementation is temporary, all the other TODO's should honestly be fine to leave in.

jamessimo reviewed (approved):
Ran locally on my system and can confirm nothing is broken or out of place. 

Absolutely a needed feature and already loving how much cleaner some XAML UI screens are looking code wise.

chavonadelal commented:
one day a holiday will come...

ps3moira commented:
Is this still being worked on?

aspiringLich commented:
Waiting on review 🤩 I have another PR for syndicate stylesheets lined up too after this

aspiringLich commented:
poke

jamessimo commented:
Are you going to fix the merge conflicts?

aspiringLich commented:
I think I'm still waiting on review from @PJB3005 ...? I should probably fix the merge conflicts soon though before they accumulate even more...

aspiringLich commented:
Let me be clear; as I see it, it's ready to merge. Its **been** ready to merge IMO. Ive just not had a maintainer tell me "yes I agree and I'll merge it after you deal with the conflicts" so I've been waiting.

sowelipililimute commented on Content.Client/Stylesheets/Colorspace/OkLabColor.cs:
```diff
@@ -0,0 +1,106 @@
+///     Oklch hue and chroma are also provided in <see cref="ColorExtensions"/>
+/// </summary>
+[PublicAPI]
+public struct OklabColor
```
we have this in robust toolbox now as of https://github.com/space-wizards/RobustToolbox/pull/6067

sowelipililimute commented on Content.Client/Stylesheets/Palette/ColorPalette.cs:
```diff
@@ -0,0 +1,73 @@
+/// <remarks>
+///     Don't be afraid to add a lot of fields here! This class is made for readability.
+/// </remarks>
+public sealed class ColorPalette(string hex = "#000000")
```
the way this is constructed feels like a weird pattern, what's stopping this from just having a static method or constructor with a bunch of optional values that get `Thing ??= ComputeFrom(Base)`ed

aspiringLich replied:
Huh. Didn't know [that](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments) was a thing. C# is such a silly "programming" ""language"". It is a lot cleaner now.

sowelipililimute commented on Content.Client/Stylesheets/StyleNano.cs:
```diff
@@ -41,6 +41,8 @@ public static Font NotoStack(this IResourceCache resCache, string variation = "R
 
     }
     // STLYE SHEETS WERE A MISTAKE. KILL ALL OF THIS WITH FIRE
+    [Obsolete("Please use the new sheetlet system to define styles, and remove all references to this class as it may be deleted in the future")]
+    // i did :)
```
if you're going to be porting every single UI to the new sheetlets in this PR, i don't see why we're hanging onto StyleNano?

aspiringLich replied:
I figured downstream forks would want to keep StyleNano around because while our UI's don't reference it anymore, they probably have UI's that do and I don't want to force them to port their code over if I don't have to. Give them some time to adjust and all that.

sowelipililimute commented on Content.Client/Stylesheets/StyleSpace.cs:
```diff
@@ -8,6 +8,7 @@
 
 namespace Content.Client.Stylesheets
 {
+    [Obsolete("Please use the new sheetlet system to define styles, and remove all references to this class as it may be deleted in the future")]
     public sealed class StyleSpace : StyleBase
```
similar question as for StyleNano

sowelipililimute reviewed (changes requested):
i'd definitely be interested in seeing this rebased onto new stuff, we have some new UIs + RT will be getting OKLAB/LCH support in its next release

the bulk of what I see here is nice, I have some questions about why we're retaining the old StyleNano/Space whilst also porting hundreds of UIs to the new stuff. is the porting not comprehensive, or...?

also, lots of TODOs strewn throughout new code that should be addressed or removed, or explicitly indicated as for being for a later PR

aspiringLich commented:
Oh? Is it finally hapenning?

I'm currently on vacation but I will work on this soon enough. Like within the next week.

aspiringLich commented:
Also the TODO's are there because when I was transferring the styles over from StyleNano I left TODO's anywhere where the rule wasn't broken but hardcoded something or did something that wasn't trivial to translate over. I think this PR changes enough 😅, so that can be for a future PR.

aspiringLich commented on Content.Client/Stylesheets/BaseStylesheet.cs:
```diff
@@ -0,0 +1,32 @@
+    /// <remarks>
+    ///     Must be a constant, changes to this after construction will not be reflected.
+    /// </remarks>
+    public virtual Type StylesheetConfigType => typeof(NoConfig);
```
That was part of the original PR. I think the intention was to allow you to provide colors or styling or whatever to change the geneee re ated stylesheet so the user could provide their own fonts or colors or something. I'll put in a todo.

FairlySadPanda commented:
I've given this a sanity check after integrating against current master. There's a few minor visual alterations, I think most notably for the chat window, but I didn't spot anything broken.

Stuff I checked:
1. The windows aghosts can pull up
2. Options menus
3. Lobby menus
4. Admin tools
5. The PDA, including being able to spawn a Nanotask task and being able to send a news message and be able to read it on the PDA.

Spooky PR but as far as I can see there are no outstanding changes requested on it.

sowelipililimute reviewed (changes requested):
Codewise, this looks fine, but it's time to interrogate the noticeable visual regressions:
- Options dropdowns lost their background, which makes them hard to read as a distinct menu
  <img width="458" height="268" alt="grafik" src="https://github.com/user-attachments/assets/6f2b016b-3ec7-4c20-8c27-8e9f41083461" />
- The contrast between disabled and active buttons is considerably worse
  <img width="836" height="431" alt="grafik" src="https://github.com/user-attachments/assets/a622759c-49e9-4c36-98b7-54b4c37d7f80" />
  This is most blatant in the lathe menus and vending machines, where an out-of-stock thing looks more enabled than the actually enabled things
  <img width="796" height="540" alt="grafik" src="https://github.com/user-attachments/assets/ea3ad91b-38dc-44b2-9bd0-aad4f5837e75" />
- Selected list items are indistinguishable from the list background
  <img width="785" height="540" alt="grafik" src="https://github.com/user-attachments/assets/5a7e43c1-4416-4026-96cf-f6ccda9cd97e" />
  <img width="293" height="369" alt="grafik" src="https://github.com/user-attachments/assets/aea7b430-9563-486d-8444-c77ba9f3bc3e" />
- The change to make pressed buttons darken instead of green is a pretty substantial change that doesn't mesh with a lot of UIs, e.g. access configurator where the distinction between various button states is poorly telegraphed
  <img width="862" height="365" alt="grafik" src="https://github.com/user-attachments/assets/46151c9f-7b36-4766-b123-2d33c7a121e9" />
- Destructive buttons are no longer red
  <img width="290" height="270" alt="grafik" src="https://github.com/user-attachments/assets/6132f557-7f54-4832-a613-7b8995be58fe" />
  <img width="379" height="459" alt="grafik" src="https://github.com/user-attachments/assets/8663c69b-6cd9-497a-a47e-719fc3099678" />
- `LabelKeyText` seems to not be working
  <img width="473" height="150" alt="grafik" src="https://github.com/user-attachments/assets/510f6a4e-f0be-4b6f-ad6b-0bf5fb188c25" />

Not strictly an issue, but the chat window being substantially darker is Very noticeable
<img width="538" height="1031" alt="grafik" src="https://github.com/user-attachments/assets/05a2a8d2-b4a8-48ef-9c58-a7497b075e5a" />

EmoGarbage404 commented on Content.Client/Stylesheets/StyleNano.cs:
```diff
@@ -515,10 +530,10 @@ public StyleNano(IResourceCache resCache) : base(resCache)
 
-            var squareTex = resCache.GetTexture("/Textures/Interface/Nano/square.png");
-            var listContainerButton = new StyleBoxTexture
+            // var squareTex = resCache.GetTexture("/Textures/Interface/Nano/square.png");
```
commented out code here. unsure what this is for.

EmoGarbage404 commented on Content.Client/UserInterface/Systems/Chat/ChatUIController.cs:
```diff
@@ -278,7 +278,7 @@ private void SetChatWindowOpacity(float opacity)
             color = propStyleBoxFlat.BackgroundColor;
         else
-            color = StyleNano.ChatBackgroundColor;
+            color = Color.Black;
```
still needs to be changed back. just leaving for perpetuity.

sowelipililimute commented:
FWIW, I intend to get this through the final stretch, anything not trivially implementable will need to be done w/ followup PRs

sowelipililimute commented on Content.Client/UserInterface/Controls/MenuButton.cs:
```diff
@@ -13,17 +13,18 @@ public sealed class MenuButton : ContainerButton
-    private static readonly Color ColorHovered = Color.FromHex("#9699bb");
-    private static readonly Color ColorRedHovered = Color.FromHex("#FFFFFF");
-    private static readonly Color ColorPressed = Color.FromHex("#789B8C");
+    // TODO: KIIIIIILLLLLLLLLLLLLLLLLLLLLLLLLLL --kaylie.
```
removed dead code, but the styling can be done in a follow up

sowelipililimute commented:
stylesheetcomponent was unimplemented, it can be done in a follow-up PR

aspiringLich commented:
It's like, in an acceptable state.  I have no major qualms with merging this with whatever minor adjustments they make.

Simyon264 commented:
I cleared it with PJB, is fine to merge without her approval.

perryprog commented:
Mfw one week has passed:

🥳
